File: packages/sapling-ios/Sources/Sapling/Sapling.swift
import SaplingFFI

public struct Sapling {
    
    public init() {}
    
    // MARK: Authorizing Key

    public func getProofAuthorizingKey(from spendingKey: ExtendedSpendingKey) throws -> ProofAuthorizingKey {
        var keyCount = 0
        guard let key = c_pak_from_xsk(spendingKey, spendingKey.count, &keyCount)?.toArray(count: keyCount) else {
            throw Error.deriveKeyFailed
        }

        return key
    }

    // MARK: Commitment

    public func verifyCommitment(_ commitment: [UInt8], to address: Address, forValue value: UInt64, with rcm: Rcm) throws -> Bool {
        var cmuCount = 0
        guard let cmu = c_compute_cmu(address, address.count, value, rcm, rcm.count, &cmuCount)?.toArray(count: cmuCount) else {
            throw Error.computeCommitmentFailed
        }

        return commitment == cmu
    }

    // MARK: Init

    public func initParameters(spend spendParameters: [UInt8], output outputParameters: [UInt8]) throws {
        guard c_init_params(spendParameters, spendParameters.count, outputParameters, outputParameters.count) else {
            throw Error.initParametersFailed
        }
    }

    // MARK: Key Agreement

    public func keyAgreement(p: [UInt8], sk: [UInt8]) throws -> [UInt8] {
        var kaCount = 0
        guard let ka = c_key_agreement(p, p.count, sk, sk.count, &kaCount)?.toArray(count: kaCount) else {
            throw Error.createKeyAgreementFailed
        }

        return ka
    }

    // MARK: Merkle Tree

    public func merkleHash(ofDepth depth: Int, lhs: [UInt8], rhs: [UInt8]) throws -> [UInt8] {
        var merkleHashCount = 0
        guard let merkleHash = c_merkle_hash(depth, lhs, lhs.count, rhs, rhs.count, &merkleHashCount)?.toArray(count: merkleHashCount) else {
            throw Error.createMerkleHashFailed
        }

        return merkleHash
    }

    // MARK: Nullifier

    public func computeNullifier(using viewingKey: ExtendedFullViewingKey, to address: Address, forValue value: UInt64, with rcm: Rcm, at position: UInt64) throws -> [UInt8] {
        var nullifierCount = 0
        guard let nullifier = c_compute_nullifier_with_xfvk(
                viewingKey, viewingKey.count,
                address, address.count,
                value,
                rcm, rcm.count,
                position,
                &nullifierCount
        )?.toArray(count: nullifierCount) else {
            throw Error.computeNullifierFailed
        }

        return nullifier
    }

    // MARK: Output Description

    public func prepareOutputDescription(
        with context: Context,
        using viewingKey: ExtendedFullViewingKey,
        to address: Address,
        withRcm rcm: Rcm,
        ofValue value: UInt64
    ) throws -> [UInt8] {
        var descriptionCount = 0
        guard let description = c_output_description_from_xfvk(context, viewingKey, viewingKey.count, address, address.count, rcm, rcm.count, value, &descriptionCount)?.toArray(count: descriptionCount) else {
            throw Error.outputDescriptionFailed
        }

        return description
    }

    public func preparePartialOutputDescription(
        with context: Context,
        to address: Address,
        withRcm rcm: Rcm,
        withEsk esk: Esk,
        ofValue value: UInt64
    ) throws -> [UInt8] {
        var descriptionCount = 0
        guard let description = c_partial_output_description(
                context,
                address, address.count,
                rcm, rcm.count,
                esk, esk.count,
                value,
                &descriptionCount
        )?.toArray(count: descriptionCount) else {
            throw Error.outputDescriptionFailed
        }

        return description
    }

    // MARK: Payment Address

    public func getPaymentAddress(from viewingKey: ExtendedFullViewingKey, at index: Index) throws -> [UInt8] {
        var addressCount = 0
        guard let address = c_payment_address_from_xfvk(viewingKey, viewingKey.count, index, index.count, &addressCount)?.toArray(count: addressCount) else {
            throw Error.createAddressFailed
        }

        return address
    }

    public func getNextPaymentAddress(from viewingKey: ExtendedFullViewingKey, lastAt index: Index) throws -> [UInt8] {
        var addressCount = 0
        guard let address = c_next_payment_address_from_xfvk(viewingKey, viewingKey.count, index, index.count, &addressCount)?.toArray(count: addressCount) else {
            throw Error.createAddressFailed
        }

        return address
    }

    public func getRawPaymentAddress(from incomingViewingKey: IncomingViewingKey, with diversifier: Diversifier) throws -> [UInt8] {
        var addressCount = 0
        guard let address = c_payment_address_from_ivk(
                incomingViewingKey, incomingViewingKey.count,
                diversifier, diversifier.count,
                &addressCount
        )?.toArray(count: addressCount) else {
            throw Error.createAddressFailed
        }

        return address
    }

    public func getDiversifier(fromRaw rawAddress: Address) throws -> [UInt8] {
        var diversifierCount = 0
        guard let diversifier = c_diversifier_from_payment_address(rawAddress, rawAddress.count, &diversifierCount)?.toArray(count: diversifierCount) else {
            throw Error.extractFromAddressFailed
        }

        return diversifier
    }

    public func getPkd(fromRaw rawAddress: Address) throws -> [UInt8] {
        var pkdCount = 0
        guard let pkd = c_pkd_from_payment_address(rawAddress, rawAddress.count, &pkdCount)?.toArray(count: pkdCount) else {
            throw Error.extractFromAddressFailed
        }

        return pkd
    }

    // MARK: Proving Context

    public func initProvingContext() throws -> Context {
        guard let context = c_init_proving_context() else {
            throw Error.initProvingContextFailed
        }

        return context
    }

    public func dropProvingContext(_ context: Context) {
        c_drop_proving_context(context)
    }

    // MARK: Rand

    public func randR() throws -> [UInt8] {
        var rCount = 0
        guard let r = c_rand_r(&rCount)?.toArray(count: rCount) else {
            throw Error.createRandomScalarFailed
        }

        return r
    }

    // MARK: Signature

    public func createBindingSignature(with context: Context, balance: Int64, sighash: [UInt8]) throws -> [UInt8] {
        var signatureCount = 0
        guard let signature = c_binding_signature(context, balance, sighash, sighash.count, &signatureCount)?.toArray(count: signatureCount) else {
            throw Error.createBindingSignatureFailed
        }

        return signature
    }

    // MARK: Spend Description

    public func prepareSpendDescriptionWithSpendingKey(
        with context: Context,
        using spendingKey: ExtendedSpendingKey,
        to address: Address,
        withRcm rcm: Rcm,
        withAr ar: Ar,
        ofValue value: UInt64,
        withAnchor anchor: Anchor,
        at merklePath: MerklePath
    ) throws -> [UInt8] {
        var descriptionCount = 0
        guard let description = c_spend_description_from_xsk(
                context,
                spendingKey, spendingKey.count,
                address, address.count,
                rcm, rcm.count,
                ar, ar.count,
                value,
                anchor, anchor.count,
                merklePath, merklePath.count,
                &descriptionCount
        )?.toArray(count: descriptionCount) else {
            throw Error.spendDescriptionFailed
        }

        return description
    }

    public func prepareSpendDescriptionWithAuthorizingKey(
        with context: Context,
        using authorizingKey: ProofAuthorizingKey,
        to address: Address,
        withRcm rcm: Rcm,
        withAr ar: Ar,
        ofValue value: UInt64,
        withAnchor anchor: Anchor,
        at merklePath: MerklePath
    ) throws -> [UInt8] {
        var descriptionCount = 0
        guard let description = c_spend_description_from_pak(
                context,
                authorizingKey, authorizingKey.count,
                address, address.count,
                rcm, rcm.count,
                ar, ar.count,
                value,
                anchor, anchor.count,
                merklePath, merklePath.count,
                &descriptionCount
        )?.toArray(count: descriptionCount) else {
            throw Error.spendDescriptionFailed
        }

        return description
    }

    public func signSpendDescription(_ spendDescription: [UInt8], using spendingKey: ExtendedSpendingKey, with ar: Ar, sighash: [UInt8]) throws -> [UInt8] {
        var signedDescriptionCount = 0
        guard let signedDescription = c_sign_spend_description_with_xsk(
                spendDescription, spendDescription.count,
                spendingKey, spendingKey.count,
                ar, ar.count,
                sighash, sighash.count,
                &signedDescriptionCount
        )?.toArray(count: signedDescriptionCount) else {
            throw Error.spendDescriptionFailed
        }

        return signedDescription
    }

    // MARK: Spending Key

    public func getExtendedSpendingKey(from seed: Seed, derivationPath: String) throws -> ExtendedSpendingKey {
        var keyCount = 0
        return try derivationPath.withCString { derivationPath in
            guard let key = c_xsk(seed, seed.count, derivationPath, &keyCount)?.toArray(count: keyCount) else {
                throw Error.deriveKeyFailed
            }

            return key
        }
    }

    // MARK: Viewing Key

    public func getExtendedFullViewingKey(from seed: Seed, derivationPath: String) throws -> ExtendedFullViewingKey {
        var keyCount = 0
        return try derivationPath.withCString { derivationPath in
            guard let key = c_xfvk(seed, seed.count, derivationPath, &keyCount)?.toArray(count: keyCount) else {
                throw Error.deriveKeyFailed
            }

            return key
        }
    }

    public func getExtendedFullViewingKey(from spendingKey: ExtendedSpendingKey) throws -> ExtendedFullViewingKey {
        var keyCount = 0
        guard let key = c_xfvk_from_xsk(spendingKey, spendingKey.count, &keyCount)?.toArray(count: keyCount) else {
            throw Error.deriveKeyFailed
        }

        return key
    }

    public func getOutgoingViewingKey(from viewingKey: ExtendedFullViewingKey) throws -> OutgoingViewingKey {
        var keyCount = 0
        guard let key = c_ovk_from_xfvk(viewingKey, viewingKey.count, &keyCount)?.toArray(count: keyCount) else {
            throw Error.deriveKeyFailed
        }

        return key
    }

    public func getIncomingViewingKey(from viewingKey: ExtendedFullViewingKey) throws -> IncomingViewingKey {
        var keyCount = 0
        guard let key = c_xfvk_to_ivk(viewingKey, viewingKey.count, &keyCount)?.toArray(count: keyCount) else {
            throw Error.deriveKeyFailed
        }

        return key
    }
    
    public typealias Context = UnsafeMutableRawPointer
    
    public typealias Seed = [UInt8]
    public typealias ExtendedSpendingKey = [UInt8]
    public typealias ExtendedFullViewingKey = [UInt8]
    public typealias IncomingViewingKey = [UInt8]
    public typealias OutgoingViewingKey = [UInt8]
    public typealias ProofAuthorizingKey = [UInt8]
    
    public typealias Address = [UInt8]
    public typealias Diversifier = [UInt8]
    public typealias Index = [UInt8]
    
    public typealias Rcm = [UInt8]
    public typealias Esk = [UInt8]
    public typealias Ar = [UInt8]
    public typealias Anchor = [UInt8]
    public typealias MerklePath = [UInt8]
 
    public enum Error: Swift.Error {
        case computeCommitmentFailed
        case initParametersFailed
        case createKeyAgreementFailed
        case createMerkleHashFailed
        case computeNullifierFailed
        case outputDescriptionFailed
        case createAddressFailed
        case extractFromAddressFailed
        case initProvingContextFailed
        case createRandomScalarFailed
        case createBindingSignatureFailed
        case spendDescriptionFailed
        case deriveKeyFailed
    }
}

extension UnsafeMutablePointer where Pointee == UInt8 {
    func toArray(count: Int) -> [UInt8] {
        let bytes = Array(UnsafeBufferPointer(start: self, count: count))
        deinitialize(count: count)
        deallocate()
        
        return bytes
    }
}

File: packages/sapling-android/app/src/main/java/it/airgap/sapling/Sapling.kt
package it.airgap.sapling

public class Sapling {

    /******** Authorizing Key ********/

    @Throws(SaplingException::class)
    public fun getProofAuthorizingKey(spendingKey: ByteArray): ByteArray =
        extPakFromXsk(spendingKey) ?: throw SaplingException("Failed to derive proof authorizing key from spending key")

    private external fun extPakFromXsk(xsk: ByteArray): ByteArray?

    /******** Commitment ********/

    @Throws(SaplingException::class)
    public fun verifyCommitment(commitment: ByteArray, address: ByteArray, value: Long, rcm: ByteArray): Boolean {
        val cmu = extComputeCmu(address, value, rcm) ?: throw SaplingException("Failed to compute commitment")
        return commitment.contentEquals(cmu)
    }

    private external fun extComputeCmu(address: ByteArray, value: Long, rcm: ByteArray): ByteArray?

    /******** Init ********/

    @Throws(SaplingException::class)
    public fun initParameters(spendParameters: ByteArray, outputParameters: ByteArray) {
        if (!extInitParameters(spendParameters, outputParameters)) {
            throw SaplingException("Failed to initialize Sapling parameters")
        }
    }

    private external fun extInitParameters(spendParameters: ByteArray, outputParameters: ByteArray): Boolean

    /******** Key Agreement ********/

    @Throws(SaplingException::class)
    public fun keyAgreement(p: ByteArray, sk: ByteArray): ByteArray =
        extKeyAgreement(p, sk) ?: throw SaplingException("Failed to create key agreement")

    private external fun extKeyAgreement(p: ByteArray, sk: ByteArray): ByteArray?

    /******** Merkle Tree ********/

    @Throws(SaplingException::class)
    public fun merkleHash(depth: Long, lhs: ByteArray, rhs: ByteArray): ByteArray =
        extMerkleHash(depth, lhs, rhs) ?: throw SaplingException("Failed to create Merkle hash")

    private external fun extMerkleHash(depth: Long, lhs: ByteArray, rhs: ByteArray): ByteArray?

    /******** Nullifier ********/

    @Throws(SaplingException::class)
    public fun computeNullifier(viewingKey: ByteArray, address: ByteArray, value: Long, rcm: ByteArray, position: Long): ByteArray =
        extComputeNullifierWithXfvk(viewingKey, address, value, rcm, position) ?: throw SaplingException("Failed to compute nullifier")

    private external fun extComputeNullifierWithXfvk(
        xfvk: ByteArray,
        address: ByteArray,
        value: Long,
        rcm: ByteArray,
        position: Long,
    ): ByteArray?

    /******** Output Description ********/

    @Throws(SaplingException::class)
    public fun prepareOutputDescription(context: Long, viewingKey: ByteArray, address: ByteArray, rcm: ByteArray, value: Long): ByteArray =
        extOutputDescriptionFromXfvk(context, viewingKey, address, rcm, value) ?: throw SaplingException("Failed to prepare output description")

    @Throws(SaplingException::class)
    public fun preparePartialOutputDescription(context: Long, address: ByteArray, rcm: ByteArray, esk: ByteArray, value: Long): ByteArray =
        extPartialOutputDescription(context, address, rcm, esk, value) ?: throw SaplingException("Failed to prepare output description")

    private external fun extOutputDescriptionFromXfvk(
        context: Long,
        xfvk: ByteArray,
        address: ByteArray,
        rcm: ByteArray,
        value: Long,
    ): ByteArray?
    private external fun extOutputDescriptionFromXfvkWithMemo(
        context: Long,
        xfvk: ByteArray,
        address: ByteArray,
        rcm: ByteArray,
        value: Long,
        memo: ByteArray,
    ): ByteArray?
    private external fun extOutputDescriptionFromOvk(
        context: Long,
        ovk: ByteArray,
        address: ByteArray,
        rcm: ByteArray,
        value: Long,
    ): ByteArray?
    private external fun extPartialOutputDescription(context: Long, address: ByteArray, rcm: ByteArray, esk: ByteArray, value: Long): ByteArray?
    private external fun extDeriveEpkFromEsk(diversifier: ByteArray, esk: ByteArray): ByteArray?

    /******** Payment Address ********/

    @Throws(SaplingException::class)
    public fun getPaymentAddressFromViewingKey(viewingKey: ByteArray, index: ByteArray?): ByteArray =
        (if (index != null) extPaymentAddressFromXfvk(viewingKey, index)
        else extDefaultPaymentAddressFromXfvk(viewingKey))
            ?: throw SaplingException("Failed to create payment address")

    @Throws(SaplingException::class)
    public fun getNextPaymentAddressFromViewingKey(viewingKey: ByteArray, index: ByteArray): ByteArray =
        extNextPaymentAddressFromXfvk(viewingKey, index) ?: throw SaplingException("Failed to create next payment address")

    @Throws(SaplingException::class)
    public fun getRawPaymentAddressFromIncomingViewingKey(viewingKey: ByteArray, diversifier: ByteArray): ByteArray =
        extPaymentAddressFromIvk(viewingKey, diversifier) ?: throw SaplingException("Failed to create payment address")

    @Throws(SaplingException::class)
    public fun getDiversifierFromRawPaymentAddress(address: ByteArray): ByteArray =
        extDiversifierFromPaymentAddress(address) ?: throw SaplingException("Failed to extract diversifier from address")

    @Throws(SaplingException::class)
    public fun getPkdFromRawPaymentAddress(address: ByteArray): ByteArray =
        extPkdFromPaymentAddress(address) ?: throw SaplingException("Failed to extract pkd from address")

    private external fun extDefaultPaymentAddressFromXfvk(xfvk: ByteArray): ByteArray?
    private external fun extNextPaymentAddressFromXfvk(xfvk: ByteArray, index: ByteArray): ByteArray?
    private external fun extPaymentAddressFromXfvk(xfvk: ByteArray, index: ByteArray): ByteArray?
    private external fun extPaymentAddressFromIvk(ivk: ByteArray, diversifier: ByteArray): ByteArray?
    private external fun extDiversifierFromPaymentAddress(address: ByteArray): ByteArray?
    private external fun extPkdFromPaymentAddress(address: ByteArray): ByteArray?

    /******** Proving Context ********/

    public fun initProvingContext(): Long = extInitProvingContext()
    public fun dropProvingContext(context: Long) = extDropProvingContext(context)

    private external fun extInitProvingContext(): Long
    private external fun extDropProvingContext(context: Long)

    /******** Rand ********/

    @Throws(SaplingException::class)
    public fun randR(): ByteArray = extRandR() ?: throw SaplingException("Failed to create random scalar")

    private external fun extRandR(): ByteArray?

    /******** Signature ********/

    public fun createBindingSignature(context: Long, balance: Long, sighash: ByteArray): ByteArray =
        extCreateBindingSignature(context, balance, sighash) ?: throw SaplingException("Failed to create binding signature")

    private external fun extCreateBindingSignature(context: Long, balance: Long, sighash: ByteArray): ByteArray?

    /******** Spend Description ********/

    @Throws(SaplingException::class)
    public fun prepareSpendDescriptionWithSpendingKey(
        context: Long,
        spendingKey: ByteArray,
        address: ByteArray,
        rcm: ByteArray,
        ar: ByteArray,
        value: Long,
        anchor: ByteArray,
        merklePath: ByteArray,
    ): ByteArray = extSpendDescriptionFromXsk(context, spendingKey, address, rcm, ar, value, anchor, merklePath)
        ?: throw SaplingException("Failed to prepare spend description")

    @Throws(SaplingException::class)
    public fun prepareSpendDescriptionWithAuthorizingKey(
        context: Long,
        authorizingKey: ByteArray,
        address: ByteArray,
        rcm: ByteArray,
        ar: ByteArray,
        value: Long,
        anchor: ByteArray,
        merklePath: ByteArray,
    ): ByteArray = extSpendDescriptionFromPak(context, authorizingKey, address, rcm, ar, value, anchor, merklePath)
        ?: throw SaplingException("Failed to prepare spend description")

    @Throws(SaplingException::class)
    public fun signSpendDescription(spendDescription: ByteArray, spendingKey: ByteArray, ar: ByteArray, sighash: ByteArray): ByteArray =
        extSignSpendDescriptionWithXsk(spendDescription, spendingKey, ar, sighash) ?: throw SaplingException("Failed to sign spend description")

    private external fun extSpendDescriptionFromXsk(
        context: Long,
        xsk: ByteArray,
        address: ByteArray,
        rcm: ByteArray,
        ar: ByteArray,
        value: Long,
        anchor: ByteArray,
        merklePath: ByteArray,
    ): ByteArray?
    private external fun extSpendDescriptionFromPak(
        context: Long,
        pak: ByteArray,
        address: ByteArray,
        rcm: ByteArray,
        ar: ByteArray,
        value: Long,
        anchor: ByteArray,
        merklePath: ByteArray,
    ): ByteArray?
    private external fun extSignSpendDescriptionWithXsk(
        spendDescription: ByteArray,
        xsk: ByteArray,
        ar: ByteArray,
        sighash: ByteArray,
    ): ByteArray?

    /******** Spending Key ********/

    @Throws(SaplingException::class)
    public fun getExtendedSpendingKey(seed: ByteArray, derivationPath: String): ByteArray =
        extXsk(seed, derivationPath) ?: throw SaplingException("Failed to create extended spending key")

    private external fun extXsk(seed: ByteArray, derivationPath: String): ByteArray?

    /******** Viewing Key ********/

    @Throws(SaplingException::class)
    public fun getExtendedFullViewingKey(seed: ByteArray, derivationPath: String): ByteArray =
        extXfvk(seed, derivationPath) ?: throw SaplingException("Failed to create extended full viewing key")

    @Throws(SaplingException::class)
    public fun getExtendedFullViewingKeyFromSpendingKey(spendingKey: ByteArray): ByteArray =
        extXfvkFromXsk(spendingKey) ?: throw SaplingException("Failed to derive extended full viewing key from spending key")

    @Throws(SaplingException::class)
    public fun getOutgoingViewingKey(viewingKey: ByteArray): ByteArray =
        extOvkFromXfvk(viewingKey) ?: throw SaplingException("Failed to derive outgoing viewing key from extended full viewing key")

    @Throws(SaplingException::class)
    public fun getIncomingViewingKey(viewingKey: ByteArray): ByteArray =
        extXfvkToIvk(viewingKey) ?: throw SaplingException("Failed to derive incoming viewing key from extended full viewing key")

    private external fun extXfvk(seed: ByteArray, derivationPath: String): ByteArray?
    private external fun extXfvkFromXsk(xsk: ByteArray): ByteArray?
    private external fun extOvkFromXfvk(xfvk: ByteArray): ByteArray?
    private external fun extXfvkToIvk(xfvk: ByteArray): ByteArray?

    public companion object {
        init {
            System.loadLibrary("sapling")
        }
    }
}